File: master-list-netlify-v3/js/normalize.js
import { VAGUE_QUANTITY_PREFIXES, NAME_MODIFIERS, MAX_INGREDIENT_NAME_LENGTH, MAX_INGREDIENT_NAME_WORDS } from "./config.js";

// "a drizzle of sesame oil" -> "sesame oil"
export function stripVagueQuantityPrefix(raw) {
  if (typeof raw !== "string") return raw;
  let s = raw.trim();
  for (const pattern of VAGUE_QUANTITY_PREFIXES) {
    s = s.replace(pattern, "");
  }
  return s.trim();
}

// Lowercase, strip parenthetical content entirely, strip descriptor words,
// naive singularize. Returns a normalized grouping key for the ingredient.
export function normalizeIngredientName(raw) {
  if (typeof raw !== "string") return null;

  let s = raw.trim().toLowerCase();
  s = s.replace(/\([^)]*\)/g, "");
  s = s.replace(/\[[^\]]*\]/g, "");
  s = s.replace(/[/,]/g, " ");
  s = s.replace(/[^a-z\s-]/g, "");
  s = s.replace(/\s+/g, " ").trim();

  let tokens = s.split(" ").filter((t) => t && !NAME_MODIFIERS.has(t));
  s = tokens.join(" ").trim();

  const words = s.split(" ").map((w) => {
    if (w.length > 3 && w.endsWith("s") && !w.endsWith("ss")) {
      return w.slice(0, -1);
    }
    return w;
  });
  s = words.join(" ").trim();

  return s || raw.trim().toLowerCase();
}

export function isProbablyPastedText(ingredient) {
  if (typeof ingredient !== "string") return false;
  const tooLong = ingredient.length > MAX_INGREDIENT_NAME_LENGTH;
  const tooManyWords = ingredient.trim().split(/\s+/).length > MAX_INGREDIENT_NAME_WORDS;
  return tooLong || tooManyWords;
}

// Takes raw extracted rows and returns { clean, review }.
// clean rows get an added `ingredientNormalized` field.
// review rows are flagged as likely pasted recipe text, kept separately.
export function cleanAndNormalize(rows) {
  const clean = [];
  const review = [];

  for (const row of rows) {
    const stripped = stripVagueQuantityPrefix(row.ingredient);
    const updated = { ...row, ingredient: stripped };

    if (isProbablyPastedText(stripped)) {
      review.push(updated);
    } else {
      updated.ingredientNormalized = normalizeIngredientName(stripped);
      clean.push(updated);
    }
  }

  return { clean, review };
}

File: ingredient-extractor/ingredient_extractor/normalize.py
"""
Cleans raw ingredient-name strings so the same ingredient, written in
different ways, groups together (e.g. "Baby spinach" / "fresh spinach" /
"Spinach" -> "spinach"; "Kosher salt (for fries)" / "Salt" -> "salt").

Also flags rows that look like pasted recipe text rather than a real
ingredient name, so they can be routed to a manual-review pile instead of
silently corrupting the combined list.
"""

from __future__ import annotations

import re

import pandas as pd

from . import config

_VAGUE_PREFIX_RE = [re.compile(p, re.IGNORECASE) for p in config.VAGUE_QUANTITY_PREFIXES]


def strip_vague_quantity_prefix(raw: str) -> str:
    """'a drizzle of sesame oil' -> 'sesame oil'"""
    if not isinstance(raw, str):
        return raw
    s = raw.strip()
    for pattern in _VAGUE_PREFIX_RE:
        s = pattern.sub("", s)
    return s.strip()


def normalize_ingredient_name(raw: str) -> str | None:
    """Lowercase, strip parenthetical content, strip descriptor words, naive
    singularize. Returns a normalized grouping key for the ingredient."""
    if not isinstance(raw, str):
        return None

    s = raw.strip().lower()
    s = re.sub(r"\([^)]*\)", "", s)   # remove parenthetical content ENTIRELY (not just the parens)
    s = re.sub(r"\[[^\]]*\]", "", s)
    s = re.sub(r"[/,]", " ", s)
    s = re.sub(r"[^a-z\s\-]", "", s)
    s = re.sub(r"\s+", " ", s).strip()

    tokens = [t for t in s.split(" ") if t and t not in config.NAME_MODIFIERS]
    s = " ".join(tokens).strip()

    # naive singularize: strip trailing 's' unless the word ends in 'ss' or is short
    words = []
    for w in s.split(" "):
        if len(w) > 3 and w.endswith("s") and not w.endswith("ss"):
            w = w[:-1]
        words.append(w)
    s = " ".join(words).strip()

    return s if s else raw.strip().lower()


def _is_probably_pasted_text(ingredient: str) -> bool:
    if not isinstance(ingredient, str):
        return False
    too_long = len(ingredient) > config.MAX_INGREDIENT_NAME_LENGTH
    too_many_words = len(ingredient.split()) > config.MAX_INGREDIENT_NAME_WORDS
    return too_long or too_many_words


def clean_and_normalize(df: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame]:
    """Takes the raw extracted DataFrame and returns (clean_df, review_df).

    clean_df has an added `ingredient_normalized` column.
    review_df contains rows flagged as likely pasted recipe text, kept
    separately so nothing is silently dropped.
    """
    df = df.copy()
    df["ingredient"] = df["ingredient"].apply(strip_vague_quantity_prefix)

    flagged = df["ingredient"].apply(_is_probably_pasted_text)
    review_df = df[flagged].copy()
    clean_df = df[~flagged].copy()

    clean_df["ingredient_normalized"] = clean_df["ingredient"].apply(normalize_ingredient_name)

    return clean_df.reset_index(drop=True), review_df.reset_index(drop=True)
